astronomy/orbit_ground_track_body.hpp

```diff
@@ -0,0 +1,164 @@
+﻿#pragma once
+
+#include "astronomy/orbit_ground_track.hpp"
+
+#include <limits>
+
+#include "physics/apsides.hpp"
+
+namespace principia {
+namespace astronomy {
+namespace internal_orbit_ground_track {
+
+using geometry::Position;
+using geometry::Vector;
+using physics::ComputeNodes;
+using quantities::Mod;
+using quantities::UnwindFrom;
+using quantities::si::Radian;
+
+// Note that the origin of this celestial longitude is arbitrary: it is not the
+// node of the orbit around the sun (the equinox).  If |PrimaryCentred| is the
+// GCRS, this is the right ascension (with respect to the mean equinox of
+// J2000).
+template<typename PrimaryCentred>
+Angle CelestialLongitude(Position<PrimaryCentred> const& q) {
+  // TODO(egg): |ToSpherical| is wasteful, as we discard the latitude.
+  return (q - PrimaryCentred::origin).coordinates().ToSpherical().longitude;
+}
+
+// The resulting angle is not normalized.
+template<typename Iterator, typename Inertial>
+Angle PlanetocentricLongitude(Iterator const& it,
+                              RotatingBody<Inertial> const& primary) {
+  return CelestialLongitude(it.degrees_of_freedom().position()) -
+         primary.AngleAt(it.time()) - π / 2 * Radian;
+}
+
+// The resulting angle is not normalized.
+template<typename Iterator>
+Angle MeanSolarTime(Iterator const& it,
+                    OrbitGroundTrack::MeanSun const& mean_sun) {
+  Time const t = it.time() - mean_sun.epoch;
+  return π * Radian + CelestialLongitude(it.degrees_of_freedom().position()) -
+         (mean_sun.mean_longitude_at_epoch +
+          (2 * π * Radian * t / mean_sun.year));
+}
+
+template<typename PrimaryCentred>
+Interval<Angle> MeanSolarTimesOfNodes(
+    DiscreteTrajectory<PrimaryCentred> const& nodes,
+    OrbitGroundTrack::MeanSun const& mean_sun) {
+  Interval<Angle> mean_solar_times;
+  std::optional<Angle> mean_solar_time;
+  for (auto node = nodes.Begin(); node != nodes.End(); ++node) {
+    if (mean_solar_time.has_value()) {
+      mean_solar_time =
+          UnwindFrom(*mean_solar_time, MeanSolarTime(node, mean_sun));
+    } else {
+      mean_solar_time = Mod(MeanSolarTime(node, mean_sun), 2 * π * Radian);
+    }
+    mean_solar_times.Include(*mean_solar_time);
+  }
+  return mean_solar_times;
+}
+
+template<typename PrimaryCentred, typename Inertial>
+Interval<Angle> ReducedLongitudesOfEquatorialCrossings(
+    DiscreteTrajectory<PrimaryCentred> const& nodes,
+    RotatingBody<Inertial> const& primary,
+    OrbitRecurrence const& nominal_recurrence,
+    std::optional<Angle>& initial_offset) {
+  Interval<Angle> reduced_longitudes;
+  int n = 0;
```

Comment:
This is only used to compute the equatorial shift, you could do that by addition through the loop, that would be more elegant.